Support other forms of quote node in allocated macro function elision

vtjnash opened the pull request:
Refs #58057

vtjnash committed:
Support kwargs, splatting, and other forms of quote node in allocated macro optimization

Refs #58057

vtjnash commented:
Yes, but you're still not allowed to have highly complicated expressions with arbitrary side-effects (`MOI.Nonlinear.eval_multivariate_hessian`) to benefit from this automatic simple-call simplification

odow commented:
> but you're still not allowed to have highly complicated expressions with arbitrary side-effects

I'm not sure I understand this part. The code does not allocate if I remove the `@allocated` macro. Previously, `@allocated` reported `0`. Now it reports `> 0`? That seems like a break in behavior, even if the fix is intended to allow optimizations for other code.

JeffBezanson commented:
Maybe we should just make `@allocated` officially only operate on the called function, i.e. allocations in argument expressions are not counted. That's a more dramatic change but it's probably what is wanted most of the time.

vtjnash commented:
Perhaps most of the time, but it would be a breaking change for several packages

vtjnash committed:
fix

JeffBezanson commented:
Fair enough. But the unintuitive thing here is that `A.B.C` is seen as a name, and not a computation. If you switch it to
```
                  f = MOI.Nonlinear.eval_multivariate_hessian
                  ret = @my_allocated f(r, :^, H, x)
```
you get 0 as expected. Maybe the right heuristic is just "is the expression a call"? Then if you really want to measure everything, all you need to do is wrap it in begin...end.

vtjnash committed:
switch to non-call-based closure capture optimization

Cause it to capture all values of any expression (even values of
globals), not just values passed to a single call.

vtjnash commented:
Shoot, if it needs a test and all that, this PR isn't worth it anymore. There's too many cases in the wild already doing various things. The PR currently breaks because we don't know that things are globals, and capturing those as arguments fails very badly for inference if they turn out to have been a Module.
